add docs (#10472)
hashicorp/packer: website/content/guides/1.7-plugin-upgrade/index.mdx

---
page_title: Upgrading your plugin to use the Packer plugin sdk
sidebar_title: Upgrade Your Plugin to use the Packer plugin sdk
---

> # Upgrading your plugin to be compatible with Packer v1.7.0

Author: I'll add a note along these lines now.

## Update plugins to use the new SDK

### Why update to the new SDK?

The goal of the SDK is to clearly separate the Packer core from the Packer plugins; as a plugin maintainer, you should only have to import the SDK and not the core. The SDK will allow us to use semantic versioning to express the changes to the API that our maintainers are expected to use, and will help us keep a clearer promise to you about the location and functionality of our helper tools.

Packer is currently backwards compatible with the old API because the plugin API itself has not changed yet, so if you choose to continue using the old Packer imports, you will be able to do so until you need to upgrade the version of Packer you are importing to create your plugin.

### How to update plugins to use the packer-plugin-SDK

In a best-case scenario, all you'll have to do is replace all the places you import

github.com/hashicorp/packer with github.com/hashicorp/packer-plugin-sdk.

But some of the import paths have changed more than that because we've refactored the SDK some to make it easier to discover and use helpful modules. Here are a few common paths below:

> github.com/hashicorp/packer-plugin-sdk/template/config => github.com/hashicorp/packer-plugin-sdk/template/config

Reviewer: ```suggestion
github.com/hashicorp/packer/template/config => github.com/hashicorp/packer-plugin-sdk/template/config
```

github.com/hashicorp/packer/common/random => github.com/hashicorp/packer-plugin-sdk/random
github.com/hashicorp/packer/common => github.com/hashicorp/packer-plugin-sdk/retry
github.com/hashicorp/packer/packer/rpc => github.com/hashicorp/packer-plugin-sdk/rpc
github.com/hashicorp/packer/communicator => github.com/hashicorp/packer-plugin-sdk/sdk-internals/communicator
github.com/hashicorp/packer/common/shell => github.com/hashicorp/packer-plugin-sdk/shell
github.com/hashicorp/packer/common/shell-local => github.com/hashicorp/packer-plugin-sdk/shell-local
github.com/hashicorp/packer/helper/builder/localexec => github.com/hashicorp/packer-plugin-sdk/shell-local/localexec
github.com/hashicorp/packer/common/shutdowncommand => github.com/hashicorp/packer-plugin-sdk/shutdowncommand
github.com/hashicorp/packer/template => github.com/hashicorp/packer-plugin-sdk/template
github.com/hashicorp/packer/packer/tmp => github.com/hashicorp/packer-plugin-sdk/tmp
github.com/hashicorp/packer/helper/useragent => github.com/hashicorp/packer-plugin-sdk/useragent
github.com/hashicorp/packer/common/uuid => github.com/hashicorp/packer-plugin-sdk/uuid

> **new** github.com/hashicorp/packer-plugin-sdk/version

Reviewer: Nice!!

Reviewer: If not too handy, I think this will look better inside of a table. I looked at the UI and it looks a bit confusing to look at and get information from it.

When in doubt you can search the packer-plugin-sdk repo for the name of your imported structs or functions to see where they exist in the sdk now; we have not changed struct or function names.

## Upgrade plugins to use new multi-plugin rpc server

### Why upgrade to the new rpc server?

The goal of this server is to enable multiple related plugins, for example plugins that share a hypervisor or cloud, to rely on shared libraries without having to submodule common code. The helps maintainers who are experts in a specific technology to focus on that technology without having to maintain several repositories. Having a single global plugin wrapper also helps the Packer core register, download, and install plugins. Finally, it allows more clarity in the plugin binary naming and name-usage within a template.

### Do I have to upgrade?

> For now, you do to have to upgrade your plugin. Packer will continue to understand how to discover and launch old-style plugins. However, you will not be able to register your plugin with HashiCorp until you have upgraded the plugin.

Reviewer: I was expecting a not here
```suggestion
For now, you do not have to upgrade your plugin, but we recommend doing it. Packer will continue to understand how to discover and launch old-style plugins. However, you will not be able to register your plugin with HashiCorp until you have upgraded the plugin.
```

Author: I mean init.

### How to upgrade the plugin

> We've created a scaffolding repository over at https://github.com/hashicorp/packer-plugin-scaffolding/tree/main/builder/scaffolding

Reviewer 2: Did you call out the builder scaffolding specifically to illustrate your example below? If so I made a small suggestion to the following text to call out the builder plugin directly. If this was indeed the intent I suggest changing this to

```suggestion
We've created a scaffolding repository over at https://github.com/hashicorp/packer-plugin-scaffolding/
```

> If you already have a working plugin, this should look pretty familiar. You'll notice that there are some changes to main.go.

Reviewer 2: ```suggestion
If you already have a working plugin, the [builder plugin scaffolding](https://github.com/hashicorp/packer-plugin-scaffolding/tree/main/builder/scaffolding) should look pretty familiar. You'll notice that there are some changes to main.go.
```
